test/py/test_device_get_region_info.py

```diff
@@ -0,0 +1,223 @@
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_GET_REGION_INFO, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+def test_device_get_region_info_bad_index():
+
+    payload = struct.pack("IIIIQQ", 32, 0, VFU_PCI_DEV_NUM_REGIONS, 0, 0, 0)
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_GET_REGION_INFO, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    get_reply(sock, expect=errno.EINVAL)
+
+def test_device_get_region_info_larger_argsz():
+
+    payload = struct.pack("IIIIQQ", 32 + 8, 0, VFU_PCI_DEV_BAR1_REGION_IDX,
+                          0, 0, 0)
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_GET_REGION_INFO, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    result = get_reply(sock)
+
+    assert(len(result) == 32 + 8)
+
+    info, _ = vfio_region_info(result)
+
+    assert info.argsz == 32
+    assert info.flags == (VFIO_REGION_INFO_FLAG_READ |
+                          VFIO_REGION_INFO_FLAG_WRITE)
+    assert info.index == VFU_PCI_DEV_BAR1_REGION_IDX
+    assert info.cap_off == 0
+    assert info.size == 4096
+    assert info.offset == vfu_region_to_offset(VFU_PCI_DEV_BAR1_REGION_IDX)
+
+def test_device_get_region_info_small_argsz_caps():
+    global sock
+
+    payload = struct.pack("IIIIQQ", 32, 0, VFU_PCI_DEV_BAR2_REGION_IDX, 0, 0, 0)
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_GET_REGION_INFO, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    result = get_reply(sock)
+
+    info, _ = vfio_region_info(result)
+
+    assert info.argsz == 80
+    assert info.flags == (VFIO_REGION_INFO_FLAG_READ |
+                          VFIO_REGION_INFO_FLAG_WRITE |
+                          VFIO_REGION_INFO_FLAG_MMAP |
+                          VFIO_REGION_INFO_FLAG_CAPS)
+    assert info.index == VFU_PCI_DEV_BAR2_REGION_IDX
+    assert info.cap_off == 0
+    assert info.size == 0x10000
+    assert info.offset == vfu_region_to_offset(VFU_PCI_DEV_BAR2_REGION_IDX)
+
+    # skip reading the SCM_RIGHTS
+    disconnect_client(ctx, sock)
+
+def test_device_get_region_info_caps():
+    global sock
+
+    sock = connect_client(ctx)
+
+    payload = struct.pack("IIIIQQ", 80, 0, VFU_PCI_DEV_BAR2_REGION_IDX, 0, 0, 0)
+    payload += b'\0' * (80 - 32)
+
+    hdr = vfio_user_header(VFIO_USER_DEVICE_GET_REGION_INFO, size=len(payload))
+    sock.send(hdr + payload)
+    vfu_run_ctx(ctx)
+    result = get_reply(sock)
+
+    info, result = vfio_region_info(result)
+    cap, result = vfio_region_info_cap_sparse_mmap(result)
+    area1, result = vfio_region_sparse_mmap_area(result)
+    area2, result = vfio_region_sparse_mmap_area(result)
+
+    assert info.argsz == 80
+    assert info.cap_off == 32
+
+    assert cap.id == VFIO_REGION_INFO_CAP_SPARSE_MMAP
+    assert cap.version == 1
+    assert cap.next == 0
+    assert cap.nr_areas == 2
+
+    assert area1.offset == 0x2000
```

Comment:
I suppose this assumes that `test_device_get_region_info_setup` will have run first. Maybe introduce a setup function or add decorators so that each test is independent?

Reply:
All of these individual files require the relevant _setup() to have run first.
By default, pytest runs the tests in serial order as written, which is exactly what we need.
Adding decorators for each individual test would involve a ton more scaffolding, and I don't really see the need - outside of valgrind, the `make pytest` is not slow at all.